server/src/graql/gremlin/NodesUtil.java

```diff
@@ -130,4 +144,58 @@ private static void buildDependenciesBetweenNodes(Set<Fragment> allFragments, Ma
         return subplan;
     }
 
+
+    /**
+     * We propagate across ISA edges any labels that might be found on one side of the ISA to
+     * the node on the other side of the ISA (moving the label from the label fragment to the instance node)
+     * @param parentToChild a mapping that represents the Query as a Tree (after the arborescence step)
+     */
+    static void propagateLabels(Map<Node, Set<Node>> parentToChild) {
+        /*
+        1. find Edge type nodes, filter these to the set of these that are ISA edges
+        2. for each ISA edge, find its parent, then children. if the parent has a LabelFragment -> to children
+            if the children contains a LabelFragment -> propagate to parent
+         */
+
+        // filter to nodes representing edges in the query planner
+        // then reduce to the ones that are pointing to an ISA either as parent or child (this information is stored
+        // on the middle node ID type)
+        Set<Node> isaEdgeNodes = parentToChild.keySet().stream()
+                .filter(node -> node instanceof EdgeNode)
+                .filter(node -> node.getNodeId().nodeIdType().equals(NodeId.NodeIdType.ISA))
+                .collect(Collectors.toSet());
+
+        // for each of the ISA edge nodes, find the parent and and children
+        for (Node node : isaEdgeNodes) {
+            Set<Node> children = parentToChild.get(node);
+            // can only be one parent - it's a  tree
+            Node parent = parentToChild.entrySet().stream().
+                    filter(entry -> entry.getValue().contains(node))
+                    .map(Map.Entry::getKey)
+                    .findFirst()
+                    .orElseThrow(() ->  GraqlSemanticException.create("QueryPlanner node: " + node.toString() + " has no parent"));
+
+            Set<Fragment> parentFragments = parent.getFragmentsWithoutDependency();
+            LabelFragment parentLabelFragment = (LabelFragment)parentFragments.stream().filter(f -> f instanceof LabelFragment).findFirst().orElse(null);
+            if (parentLabelFragment != null) {
+                // propagate the label to the children
+                Label label = parentLabelFragment.labels().iterator().next();
+                children.stream().filter(childNode -> childNode instanceof InstanceNode).forEach(child -> ((InstanceNode) child).setInstanceLabel(label));
```

Comment:
just FYI, to filter for type you can also do `.filter(InstanceNode.class::isInstance)`, and to cast '.map(InstanceNode.class::cast)`. Looks a bit neater in my mind.

Reply:
Just did something like this with marco! Explicitly casting now rather than ::cast however